raft/raft.go

```diff
@@ -0,0 +1,292 @@
+	log.WithFields(log.Fields{}).Infof("do follower action start")
+
+	for {
+		select {
+		case <-r.electionTimer.C:
+			// be a candidate
+			// TODO: fix race condition possibilities,
+			// should not happen because the only thread that modifies the state.role is this thread
+			r.state.ChangeRole(model.RaftRoleCandidate)
+			return
+		case <-r.appendEntriesSignal:
+			log.Debug("heartbeat received")
+			// reset election timeout
+			r.resetElectionTimer()
+			continue
+		}
+	}
+}
+
+func (r *defaultRaft) doCandidateAction() {
+	log.WithFields(log.Fields{}).Infof("do candidate action startzzz")
+	electionChan := make(chan struct{}, 1)
+	electionChan <- struct{}{}
+
+	log.WithFields(log.Fields{}).Infof("do candidate action send election")
+
+	for {
+		select {
+		case <-r.electionTimer.C:
+			log.WithFields(log.Fields{}).Infof("reset election timer")
+			// reset election timer
+			r.resetElectionTimer()
+			// do another election
+			electionChan <- struct{}{}
+		case <-r.appendEntriesSignal:
+			log.WithFields(log.Fields{}).Infof("heartbeat received")
+			r.state.Lock()
+			r.state.ChangeRole(model.RaftRoleFollower)
+			r.state.VotedFor = ""
+			r.state.Unlock()
+			return
+		case <-electionChan:
+			log.WithFields(log.Fields{}).Infof("do election")
+			shouldReturn := r.doElection()
+			if shouldReturn {
+				return
+			}
+		}
+	}
+}
+
+func (r *defaultRaft) doElection() bool {
+	log.WithFields(log.Fields{}).Infof("do election start")
+
+	r.state.Lock()
+	defer r.state.Unlock()
+
+	// 1. increment current term
+	r.state.CurrentTerm++
+	// 2. vote for self
+	r.state.VotedFor = r.state.ID
+	// 3. reset election timer
+	r.resetElectionTimer()
+	// calculate majority:
+	majority := math.Ceil(float64(len(r.config.ClusterMemberAddreses)) / float64(2))
+	// 1 because we voted for ourselves
+	var votes int32 = 1
+
+	// Do voting
+	log.WithFields(log.Fields{}).Infof("do voting")
+	requestVoteReq := model.NewRequestVoteRequestDTO(
+		r.state.CurrentTerm,
+		r.state.ID,
+		int32(len(r.state.Logs)),
+		0)
+
+	voteRespChan := make(chan *model.RequestVoteResponseDTO, len(r.config.ClusterMemberAddreses))
+
+	var wg sync.WaitGroup
+	for _, memberAddress := range r.config.ClusterMemberAddreses {
+		if memberAddress != r.address {
+			wg.Add(1)
+			go func(memberAddress string) {
+				defer wg.Done()
+				resp, err := r.raftClient.RequestVote(memberAddress, requestVoteReq)
+				if err != nil {
+					log.WithFields(log.Fields{
+						"err": err,
+					}).Error("error requesting vote")
+					return
+				}
+
+				log.WithFields(log.Fields{
+					"id":      r.state.ID,
+					"address": memberAddress,
+					"resp":    resp,
+				}).Infof("do voting done")
+				voteRespChan <- resp
+			}(memberAddress)
+		}
+	}
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		voteSeen := 0
+		for voteResp := range voteRespChan {
+			voteSeen++
+			if voteResp.VoteGranted {
+				atomic.AddInt32(&votes, 1)
+			}
+			if voteSeen == len(r.config.ClusterMemberAddreses)-1 {
```

Comment:
ini buat ensure semua servernya mesti ngasih respon vote dulu? in case salah satu servernya ngasih response terlalu lama meskipun majority vote udah setuju, ini gak bakal ngeblock election process kah?

Reply:
Ah bener, harusnya nggak perlu wait until semua node ngevote dulu. thanks will fix